### PR TITLE
Add Preact root fragment implementation replacement for `replaceNode`

### LINKED ISSUE #19607: Update the Preact `replaceNode` implementation which is being deprecated.
The Preact `replaceNode` implementation that we use [here](https://github.com/forem/forem/blob/a82d94e1dbfb850570579230609671076529f1cf/app/javascript/packs/homePageFeed.jsx#L195-L200) and in other places in the application is being deprecated. 

The Warning message that we observe is:
The signature '(vnode: ComponentChild, parent: ContainerNode, replaceNode?: Element | Text): void' of 'render' is deprecated.ts(6387) index.d.ts(293, 4): The declaration was marked as deprecated.  

The Replacement Preact 10+ implementation can be found [here](https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c).

Please replace **all instances** where this is being used. Make sure that the places this is used is tested properly.

### PR DESCRIPTION
**Pull request by jeferson-sb**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Add `preact-root-fragment` implementation for Preact 10+ for rendering multiple nodes in fragments as an replacement for `replaceNode` of the render method, as explained [here](https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c#why-do-i-need-this)

## Related Tickets & Documents

- Related Issue #19607 
- Closes #19607 

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_
- [ ] Semantic HTML implemented?
- [ ] Keyboard operability supported?
- [ ] Checked with [axe DevTools](https://www.deque.com/axe/) and addressed `Critical` and `Serious` issues?
- [ ] Color contrast tested?

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/javascript/packs/articleForm.jsx [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { h, render } from 'preact';
 import { ArticleForm } from '../article-form/articleForm';
 import { Snackbar } from '../Snackbar';
 import { getUserDataAndCsrfToken } from '@utilities/getUserDataAndCsrfToken';
+import { createRootFragment } from '../shared/preact/preact-root-fragment';
```
Comment (rt4914) [inline app/javascript/packs/articleForm.jsx:5]: Reorder this import.

```
import { createRootFragment } from '../shared/preact/preact-root-fragment';
import { getUserDataAndCsrfToken } from '@utilities/getUserDataAndCsrfToken';

```

File: app/javascript/packs/base.jsx [hunk 2]
```diff
@@ -11,6 +11,7 @@ import { initializeDashboardSort } from './initializers/initializeDashboardSort'
 import { trackCreateAccountClicks } from '@utilities/ahoy/trackEvents';
 import { showWindowModal, closeWindowModal } from '@utilities/showModal';
 import * as Runtime from '@utilities/runtime';
+import { createRootFragment } from '../shared/preact/preact-root-fragment';
```
Comment (rt4914) [inline app/javascript/packs/base.jsx:14]: Ditto. Reorder these imports as per warning.

File: app/javascript/packs/homePageFeed.jsx [hunk 3]
```diff
@@ -5,6 +5,7 @@ import { Feed } from '../articles/Feed';
 import { TodaysPodcasts, PodcastEpisode } from '../podcasts';
 import { articlePropTypes } from '../common-prop-types';
 import { getUserDataAndCsrfToken } from '@utilities/getUserDataAndCsrfToken';
+import { createRootFragment } from '../shared/preact/preact-root-fragment';
```
Comment (rt4914) [inline app/javascript/packs/homePageFeed.jsx:8]: Ditto as above.

File: app/javascript/previewCards/feedPreviewCards.jsx [hunk 4]
```diff
@@ -5,6 +5,7 @@ import {
   getDropdownRepositionListener,
 } from '@utilities/dropdownUtils';
 import { request } from '@utilities/http/request';
+import { createRootFragment } from '../shared/preact/preact-root-fragment';
```
Comment (rt4914) [inline app/javascript/previewCards/feedPreviewCards.jsx:8]: Ditto as above.

File: app/javascript/shared/preact/package.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "preact-root-fragment",
+  "version": "0.2.0",
+  "main": "./preact-root-fragment.js",
+  "module": "./preact-root-fragment.js",
+  "type": "module"
+}
```
Comment (rt4914) [inline app/javascript/shared/preact/package.json:7]: EOF is missing, add one blank line at the end.

File: app/javascript/shared/preact/preact-root-fragment.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  return parent.__k = {
+    nodeType: 1,
+    parentNode: parent,
+    firstChild: replaceNode[0],
+    childNodes: replaceNode,
+    insertBefore: insert,
+    appendChild: insert,
+    removeChild: function(c) { parent.removeChild(c); }
+  };
+}
```
Comment (rt4914) [inline app/javascript/shared/preact/preact-root-fragment.js:21]: EOF is missing, add one blank line at the end.

File: app/javascript/shared/preact/preact-root-fragment.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/**
+ * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
+ *
+ * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
+ * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
+ * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
+ */
+export function createRootFragment(parent, replaceNode) {
+  replaceNode = [].concat(replaceNode);
+  var s = replaceNode[replaceNode.length-1].nextSibling;
```
Comment (rt4914) [inline app/javascript/shared/preact/preact-root-fragment.js:10]: consts = replaceNode[replaceNode.length-1].nextSibling;

**Review by rt4914 — CHANGES_REQUESTED** [review]
Thanks for contribution @jeferson-sb 
I have suggested few changes on this PR. Thanks.

File: app/javascript/shared/preact/preact-root-fragment.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  function insert(c, r) { parent.insertBefore(c, r || s); }
+  return parent.__k = {
+    nodeType: 1,
+    parentNode: parent,
+    firstChild: replaceNode[0],
+    childNodes: replaceNode,
+    insertBefore: insert,
+    appendChild: insert,
+    removeChild: function(c) { parent.removeChild(c); }
+  };
```
Comment (filleduchaos) [inline app/javascript/shared/preact/preact-root-fragment.js:20]: This implementation doesn't work; the app fails and doesn't render anything. This is why so many of the tests are failing

Screenshot showing blank home page

There is [an update/correction](https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c?permalink_comment_id=4558744#gistcomment-4558744) in the comments in the linked gist (it's in TypeScript, but should work in JavaScript at well)

**Review by filleduchaos — CHANGES_REQUESTED** [review]
Thank you for the PR ✨ We have commit hooks to auto-fix (or at least point out) some of the linter errors that are showing up here; you can take advantage of them if you set up a full dev environment (which would also help with verifying the behavior of changes before re-requesting review)

File: app/javascript/shared/preact/preact-root-fragment.js (hunk 8 above)
Comment (jeferson-sb) [reply app/javascript/shared/preact/preact-root-fragment.js]: Thanks, I did some updates to the implementation script

File: app/javascript/shared/preact/preact-root-fragment.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/**
+ * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
+ *
+ * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
+ * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
+ * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
+ */
+export function createRootFragment(parent, replaceNode) {
```
Comment (rt4914) [inline app/javascript/shared/preact/preact-root-fragment.js:8]: This file is not covered by tests.  Please add tests for this file so as to pass the code-coverage.

Let us know if you need help with tests.

**Review by rt4914 — CHANGES_REQUESTED** [review]

**Review by rt4914 — COMMENTED** [review]
Thanks @jeferson-sb 
There are some suggestions from audit/Github-actions. You can check all of them by visiting https://github.com/forem/forem/pull/19673/files and address these issues. Thanks.

File: app/javascript/shared/preact/preact-root-fragment.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      return c;
+    },
+    appendChild: (c) => {
+      parent.insertBefore(c, s);
+      return c;
+    },
+    removeChild: (c) => {
+      parent.removeChild(c);
+      return c;
+    },
```
Comment (filleduchaos) [inline app/javascript/shared/preact/preact-root-fragment.js:34]: I'd prefer to have a spec for `removeChild` as well

**Review by filleduchaos — APPROVED** [review]
Thanks again for your contribution! 🎉

**Review by rt4914 — APPROVED** [review]
Thanks a lot for your contribution. Great work @jeferson-sb

**Comment by rt4914** [comment]
@jaw6 This PR is merge ready.

**Pull request merged**